File: JS/merge.js
/**
 * Merge sort implementation in Javascript.
 * Generates and sorts a list of 10 random integers 1-100.
 * Tracks and displays # of array accesses and comparisons completed during sorting.
 * 
 * Author: Lucas Burns
 * Version: 2017-9-26
 * 
 */

const process = require('./process.js');

/**
 * Sorts a given array of numbers using Merge sort
 * @param {Array<number>} items 
 */
function sort(items){
    if (items.length === 1) {
        accesses++;
        return items;
    }

    var mid = Math.floor(items.length / 2);
    var left = sort(items.slice(0, mid));
    var right = sort(items.slice(mid, items.length));
    var result = new Array(left.length + right.length);
    result = merge(left, right, result);
    return result;
}

/**
 * Sorting helper to sort and merge lists
 * @param {Array<number>} left 
 * @param {Array<number>} right 
 * @param {Array<number>} result 
 */
function merge(left, right, result){
    var leftIdx = 0;
    var rightIdx = 0;
    for (var i = 0; i < result.length; i++){
        if (left[leftIdx] < right[rightIdx] || rightIdx === right.length){
            result[i] = left[leftIdx++];
            comparisons++;
            accesses++;
            continue;
        }
        result[i] = right[rightIdx++];
        comparisons++;
        accesses++;
    }
    return result;
}

var a = process.populate();

console.log("Original array:	        [" + a + "]");
comparisons = 0;
accesses = 0;

a = sort(a);

console.log("Sorted array:	 	[" + a + "]");
console.log("%d comparisons		%d array accesses.", comparisons, accesses);

File: python/merge.py
#!/usr/bin/env python

"""merge.py: Sorts an array of integers using merge sort"""

import process
import math



def sort(items):
	if len(items) == 1:
		return items

	mid = int(math.floor(len(items)/2.0))
	left = sort(items[0:mid])
	right = sort(items[mid:])
	items = merge(left, right)
	return items

def merge(left, right):
	l = 0
	r = 0
	result = []
	while len(left) != 0 and len(right) != 0:
		process.comparisons += 1
		process.accesses += 1
		if left[0] < right[0]:
			temp = left[0]
			result.append(temp)
			left.remove(temp)
			continue
		temp = right[0]
		result.append(temp)
		right.remove(temp)
	if len(left) == 0:
		result += right
	else:
		result += left
	return result

a = process.populate()

print("Original array:		" + str(a))

process.initCounters()
a = sort(a)

print("Sorted array: 		" + str(a))
print(str(process.comparisons) + " comparisons		" + str(process.accesses) + " array accesses")